### PR TITLE
MWPW-132087 - [Blog2Milo] Banner Block

### PR DESCRIPTION
**Pull request by jckautzmann**
- create a Milo based banner block

Test Urls:
Before: https://main--blog--adobecom.hlx.page/en/drafts/jck/banner/page-with-banner?martech=off
After: https://banner--blog--adobecom.hlx.page/en/drafts/jck/banner/page-with-banner?martech=off

Real pages:
Before: https://main--blog--adobecom.hlx.page/en/publish/2023/05/24/adobe-and-bowie-state-university-cybersecurity-internship
After: https://banner--blog--adobecom.hlx.page/en/publish/2023/05/24/adobe-and-bowie-state-university-cybersecurity-internship

### CODE REVIEW

File: scripts/scripts.js [hunk 1]
```diff
@@ -44,7 +44,75 @@ const CONFIG = {
...
+      const img = document.createElement('img');
+      img.setAttribute('src', `${pathname}?width=${br.width}&format=${ext}&optimize=medium`);
+      img.setAttribute('loading', eager ? 'eager' : 'lazy');
+      img.setAttribute('alt', alt);
+      picture.appendChild(img);
+    }
+  });
+
+  return picture;
+}
```
Comment (rgclayton) [inline scripts/scripts.js:115]: What is all of this? This is all related to the banner?

Comment (rgclayton) [reply scripts/scripts.js:115]: NM. Just noticed the call in banner.js looking for these :)

**Comment by rgclayton** [comment]
@chrischrischris Do you mind taking a look at this? Porting over the `banner` block from the blog.

File: styles/styles.css [hunk 2]
```diff
@@ -1,13 +1,79 @@
...
+  /* grid sizes */
+  --grid-container-width: 83.4%;
+}
+
+main {
+  font-size: var(--body-font-size-m);
+  line-height: var(--body-line-height);
+  margin: 64px 0 0;
+  font-weight: var(--body-font-weight);
+}
```
Comment (rgclayton) [inline styles/styles.css:38]: I don't think we want any of `main`. I think using what we get from Milo is a better option. This makes the main font look really thin.

File: styles/styles.css [hunk 3]
```diff
@@ -1,13 +1,79 @@
...
+
+
+main h1, main h2, main h3, main h4, main h5, main h6 {
+  font-weight: var(--heading-font-weight);
+  line-height: var(--heading-line-height);
+  margin-top: 48px;
+  margin-bottom: 16px;
+}
```
Comment (rgclayton) [inline styles/styles.css:46]: Same with this. I think we go with what we are getting from Milo.

File: styles/styles.css [hunk 4]
```diff
@@ -1,13 +1,79 @@
...
+
+/* links */
+
+main a:any-link {
+  color: var(--body-color);
+  line-height: var(--body-line-height);
+}
```
Comment (rgclayton) [inline styles/styles.css:53]: Same here. This makes general links blend into content which is an accessibility concern.

**Comment by rgclayton** [comment]
@auniverseaway Any thoughts/opinions? I know you had a couple concerns with the original blog implementation of this.

File: blocks/banner/banner.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  const bannerContents = document.createElement('div');
+  bannerContents.classList.add('banner-contents');
+  block.querySelectorAll('a').forEach(async (a) => {
+    if (a && a.href) {
+      // content wrapper
+      const bannerContent = document.createElement('div');
+      bannerContent.classList.add('content-wrapper');
+
+      // get response from the URL
+      const { pathname } = new URL(a);
```
Comment (chrischrischris) [inline blocks/banner/banner.js:16]: I don't know if it matters but if the url is invalid it'll throw

File: scripts/scripts.js [hunk 6]
```diff
@@ -44,7 +44,75 @@ const CONFIG = {
...
+  });
+}
+
+/**
+ * Returns a picture element with webp and fallbacks
+ * @param {string} src The image URL
+ * @param {boolean} eager load image eager
+ * @param {Array} breakpoints breakpoints and corresponding params (eg. width)
+ */
+export function createOptimizedPicture(src, alt = '', eager = false, breakpoints = [{ media: '(min-width: 400px)', width: '2000' }, { width: '750' }]) {
```
Comment (chrischrischris) [inline scripts/scripts.js:83]: nit: line length

File: scripts/scripts.js [hunk 7]
```diff
@@ -44,7 +44,75 @@ const CONFIG = {
...
+/**
+ * Normalizes all headings within a container element.
+ * @param {Element} el The container element
+ * @param {[string]]} allowedHeadings The list of allowed headings (h1 ... h6)
+ */
+export function normalizeHeadings(el, allowedHeadings) {
+  const allowed = allowedHeadings.map((h) => h.toLowerCase());
+  el.querySelectorAll('h1, h2, h3, h4, h5, h6').forEach((tag) => {
+    const h = tag.tagName.toLowerCase();
+    if (allowed.indexOf(h) === -1) {
```
Comment (chrischrischris) [inline scripts/scripts.js:58]: nit: use array.includes

**Review by chrischrischris — COMMENTED** [review]
Ideally this wouldn't even be a block but would leverage fragments since this is pretty much a specialized version of that.  I understand we need it for backwards compatibility.

You might be able to leverage the fragment.js code, but it may make it harder to get the DOM structure you want.

The changes to scripts.js maybe would be better in utils.js

**Comment by jckautzmann** [comment]
@rgclayton @chrischrischris Thx for your feedback! I addressed it in https://github.com/adobecom/blog/pull/13/commits/4c41d987a31a8db5dbbb7ba7d1bf3eff20a00dc2
Could you please have a look at it?

File: blocks/banner/banner.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+          // banner text content
+          normalizeHeadings(responseEl, ['h3']);
+          const link = responseEl.querySelector('a');
+          link.classList.add('cta-link');
+          bannerText.append(responseEl);
+
+          // appending DOM objects
+          bannerContent.append(bannerImage);
+          bannerContent.append(bannerText);
```
Comment (rgclayton) [inline blocks/banner/banner.js:49]: Element.append() can append several nodes and strings.

```suggestion
          bannerContent.append(bannerImage, bannerText);
```

File: scripts/scripts.js [hunk 9]
```diff
@@ -138,7 +139,7 @@ const { loadArea, setConfig } = await import(`${miloLibs}/utils/utils.js`);
   decorateFigure();
   decorateVideo();
   decorateQuote();
-  decorateContent();
+  await decorateContent();
```
Comment (rgclayton) [inline scripts/scripts.js:142]: Why this change?

File: scripts/scripts.js [hunk 10]
```diff
@@ -96,18 +97,18 @@ function decorateFigure() {
...
-    const para =  document.createElement('p');
+    const para = document.createElement('p');
 
     link.href = url;
     para.append(link);
-    block.insertAdjacentElement("beforebegin", para);
+    block.insertAdjacentElement('beforebegin', para);
     block.remove();
-    
+
```
Comment (rgclayton) [inline scripts/scripts.js:111]: Aye... Thank you for fixing these. My bad :(

File: scripts/scripts.js (hunk 9 above)
Comment (jckautzmann) [reply scripts/scripts.js:148]: the lint tool was giving a warning as `decorateContent` is `async`

File: blocks/banner/banner.js (hunk 8 above)
Comment (jckautzmann) [reply blocks/banner/banner.js]: nice catch! Addressed in https://github.com/adobecom/blog/pull/13/commits/45863e761cfeaaa139bac5e18c4191d9778d82ff

File: blocks/banner/banner.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    if (!a?.href) return;
+    // content wrapper
+    const bannerContent = document.createElement('div');
+    bannerContent.classList.add('content-wrapper');
+
+    try {
+      // get response from the URL
+      const { pathname } = new URL(a);
+      const path = pathname?.replace(/\.html$/, '');
+      if (path) {
```
Comment (chrischrischris) [inline blocks/banner/banner.js:19]: nit: early return

**Review by rgclayton — APPROVED** [review]
